File: library/src/main/java/de/nenick/espressomacchiato/elements/EspButton.java
package de.nenick.espressomacchiato.elements;

import android.view.View;

import org.hamcrest.Matcher;

import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.withText;

public class EspButton extends EspView {

    public static EspButton byId(int resourceId) {
        return new EspButton(resourceId);
    }

    public static EspButton byText(String text) {
        return new EspButton(withText(text));
    }

    public EspButton(int resourceId) {
        super(resourceId);
    }

    public EspButton(Matcher<View> base) {
        super(base);
    }

    public void assertTextIs(String expected) {
        findView().check(matches(withText(expected)));
    }
}

File: esp-android/src/main/java/de/nenick/espressomacchiato/widget/EspButton.kt
package de.nenick.espressomacchiato.widget

import android.view.View
import androidx.annotation.IdRes
import androidx.test.espresso.Root
import androidx.test.espresso.matcher.ViewMatchers.withId
import de.nenick.espressomacchiato.viewinteraction.ClickActions
import de.nenick.espressomacchiato.widgetinteraction.TextAssertions
import de.nenick.espressomacchiato.internals.OpenForExtensions
import de.nenick.espressomacchiato.view.EspView
import org.hamcrest.Matcher

@Suppress("UNCHECKED_CAST")
@OpenForExtensions
class EspButton(
        viewMatcher: Matcher<View>,
        rootMatchers: Matcher<Root>? = null,
        interactions: EspButton.() -> Unit = {}
) :
        EspView(viewMatcher, rootMatchers, interactions as EspView.() -> Unit),
        ClickActions,
        TextAssertions {

    constructor(@IdRes viewId: Int, interactions: EspButton.() -> Unit = {})
            : this(withId(viewId), null, interactions)

    constructor(@IdRes viewId: Int, rootMatchers: Matcher<Root>, interactions: EspButton.() -> Unit = {})
            : this(withId(viewId), rootMatchers, interactions)
}